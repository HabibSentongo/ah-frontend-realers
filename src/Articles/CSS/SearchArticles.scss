@import '../../commons/assets/variables/font.scss';
@import '../../commons/assets/variables/colors.scss';

.search-articles  {
  border-top: 1px solid $light-gray;
  min-height: 250px;
  padding-top: 10px;

  .error-box {
    margin-top: 80px;
  }

  .section-title {
    color: $primary-blue;
    font-family: $lato-font;
    font-size: 24px;
    font-weight: 100;
    margin: 20px 0;
    text-align: center;
    width: 100%;
  }
}
