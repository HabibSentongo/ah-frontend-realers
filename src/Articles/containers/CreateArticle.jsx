import {
  Form, Row, Col, Button, ProgressBar,
} from 'react-bootstrap';
import React from 'react';
import '../CSS/CreateArticle.scss';
import { connect } from 'react-redux';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faCamera } from '@fortawesome/free-solid-svg-icons';
import Loader from 'react-loader';
import PropTypes from 'prop-types';
import { storage } from '../../commons/Util/FirebaseConfig';
import postAnArticle from '../redux/actions/CreateArticleAction';
import ErrorLabel from '../../commons/components/ErrorLabel';
import SuccessLabel from '../../commons/components/SuccessLabel';
import { TITLE_REGEX, BODY_REGEX, TAGS_REGEX } from '../constants';

export class CreateArticle extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      isFieldInvalid: {
        title: null,
        body: null,
        tags: null,
      },
      articleTitle: '',
      articleBody: '',
      articleTags: '',
      uploadProgress: 0,
      uploadError: 'No Error',
      displayProgress: 'none',
      isFormValid: false,
      articlePhoto: {
        name: 'No Photo selected',
        data: null,
      },
    };
  }

  componentWillMount() {
    const token = localStorage.getItem('user_token');
    if (token === null) {
      window.location = '/login';
    }

    return true;
  }

  handleCreateArticle = (event) => {
    event.preventDefault();
    const {
      articleTitle, articleBody, articleTags, articlePhoto,
    } = this.state;
    const articleData = {
      article: {
        title: articleTitle,
        body: articleBody,
        description: articleBody,
      },
    };

    if (articleTags !== '') {
      const tags = articleTags.split(',');
      articleData.article.tagName = tags;
    }

    const { postAnArticle } = this.props;

    if (articlePhoto.data !== null) {
      this.handleImageUpload();
    } else {
      postAnArticle(articleData);
    }
  }

  handleChange = (event) => {
    const { target: { name, value, files } } = event;
    const { isFieldInvalid } = this.state;

    switch (name) {
      case 'articleTitle':
        isFieldInvalid.title = !TITLE_REGEX.test(value);
        break;
      case 'articleBody':
        isFieldInvalid.body = !BODY_REGEX.test(value);
        break;
      case 'articleTags':
        isFieldInvalid.tags = !TAGS_REGEX.test(value);
        break;
      case 'articlePhoto':
        this.setState({
          articlePhoto: {
            name: 'One Photo Added, click here again to change photo',
            data: files[0],
          },
        });
        break;
      default:
        break;
    }
    if (name !== 'articlePhoto') {
      this.setState({
        [name]: value,
      });
    }

    if ((isFieldInvalid.title === false
        && isFieldInvalid.body === false
    )) {
      this.setState({
        isFormValid: true,
      });
    } else {
      this.setState({
        isFormValid: false,
      });
    }
  }

  handleImageUpload = () => {
    const {
      articleTitle, articleBody, articleTags, articlePhoto,
    } = this.state;

    const { postAnArticle } = this.props;

    this.setState({
      displayProgress: 'block',
    });

    const imageUpLoad = storage.ref(`article-image/${articlePhoto.data.name}`).put(articlePhoto.data);
    imageUpLoad.on('state_changed',
      (snapshot) => {
        const progress = (snapshot.bytesTransferred / snapshot.totalBytes) * 100;
        this.setState({
          uploadProgress: progress,
        });
      },
      (error) => {
        this.setState({
          uploadError: error,
        });
      },
      /* istanbul ignore next */
      (complete) => {
        storage.ref('article-image').child(articlePhoto.data.name).getDownloadURL()
          .then(url => {
            const articleData = {
              article: {
                title: articleTitle,
                body: articleBody,
                description: articleBody,
                image: url,
              },
            };
            if (articleTags !== '') {
              const tags = articleTags.split(',');
              articleData.article.tagName = tags;
            }
            postAnArticle(articleData);
          });
      });
  }

  render() {
    const {
      isFieldInvalid, articlePhoto, articleTitle, articleBody, isFormValid,
      articleTags, uploadProgress, displayProgress, uploadError,
    } = this.state;
    const {
      errorOccured,
      isPosting,
      isArticlePosted,
      createdArticle,
    } = this.props;
    if (createdArticle !== undefined) {
      setTimeout(() => {
        window.location = `/article/${createdArticle.data.slug}`;
      },
      4000);
    }
    return (
      <div className="create-article">
        <h2 className="section-title">Create Article</h2>
        <Form onSubmit={this.handleCreateArticle}>
          <Form.Group controlId="formBasicTitle">
            <Form.Control
              type="text"
              required
              name="articleTitle"
              placeholder="Article Title"
              value={articleTitle}
              onChange={this.handleChange}
              isInvalid={isFieldInvalid.title} />
            {
            isFieldInvalid.title
              ? (
                <Form.Text className="text-muted" variant="danger">
                  Article title can&apos;t be empty or contain only digits or with less than 4 characters
                </Form.Text>
              )
              : (
                <Form.Text className="text-muted" />
              )
            }
          </Form.Group>
          <Form.Group controlId="exampleForm.ControlTextarea1">
            <Form.Control
              name="articleBody"
              value={articleBody}
              onChange={this.handleChange}
              as="textarea"
              rows="15"
              placeholder="Article body"
              isInvalid={isFieldInvalid.body}
              required
            />
            {
            isFieldInvalid.body
              ? (
                <Form.Text className="text-muted">
                  Article body can&apos;t not be empty, should be atleast 100 characters long
                </Form.Text>
              )
              : (
                <Form.Text className="text-muted" />
              )
            }
          </Form.Group>
          <Form.Group controlId="formBasicTag">
            <Form.Control
              name="articleTags"
              value={articleTags}
              onChange={this.handleChange}
              type="text"
              placeholder="Article Tags"
              isInvalid={isFieldInvalid.tags} />
            {
            isFieldInvalid.tags
              ? (
                <Form.Text className="text-muted">
                  Enter tags should only be letters separated by comma like books,education
                </Form.Text>
              )
              : (
                <Form.Text className="text-muted" />
              )
            }
          </Form.Group>
          <Form.Group>
            <Form.Label htmlFor="articlePhoto" style={{ cursor: 'pointer' }}>
              <h5 className="article-photo">
                <span>
                  <FontAwesomeIcon icon={faCamera} />
                  {' '}
                  Add Photo
                </span>
                <span className="article-photo-name">{articlePhoto.name}</span>
              </h5>
              <Form.Control
                    id="articlePhoto"
                    name="articlePhoto"
                    type="file"
                    accept=".png,.jpg"
                    onChange={this.handleChange}
                    style={{ display: 'none' }}
                />
            </Form.Label>
            <Row style={{ display: displayProgress }}>
              {
                uploadError === 'No Error'
                  ? (
                    <Col sm={12}>
                      <ProgressBar now={uploadProgress} label={`${uploadProgress}%`} animated />
                    </Col>
                  )
                  : (
                    <ErrorLabel
                      message={String(uploadError)}
                      tip="try uploading the image again"
                    />
                  )
              }

            </Row>
          </Form.Group>
          <Row className="status-section">
            <Col sm={12}>
              {
                errorOccured !== undefined
                && (
                <ErrorLabel
                message={String(errorOccured.errorMessage.errors.body)}
                tip="edit your article body"
                />
                )
              }
              {
                isArticlePosted && <SuccessLabel message="Article has been successfully created" />
              }
            </Col>
          </Row>
          <Row>
            <Col sm={12} className="call-for-action">
              <Loader loaded={!isPosting} color="#00a4d3">
                <Button type="submit" disabled={!isFormValid} id="submitForm">
               Create Article
                </Button>
              </Loader>
            </Col>
          </Row>
        </Form>
      </div>
    );
  }
}

CreateArticle.protoTypes = {
  postAnArticle: PropTypes.func.isRequired,
  isArticlePosted: PropTypes.bool.isRequired,
};

export const mapStateToProps = state => ({
  isPosting: state.createArticle.isPosting,
  errorOccured: state.createArticle.error,
  isArticlePosted: state.createArticle.status,
  createdArticle: state.createArticle.item,
});

export default connect(mapStateToProps, { postAnArticle })(CreateArticle);
