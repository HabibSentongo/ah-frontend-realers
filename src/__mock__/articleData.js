/* eslint-disable import/prefer-default-export */
export const ARTICLE_DATA = [
  {
    title: 'string',
    description: 'string',
    body: 'strinfghjhhg',
    createdAt: '2019-06-19T09:16:23.749462Z',
    updatedAt: '2019-06-19T09:16:23.749489Z',
    author: {
      email: 'cedric.lusiba@andela.com',
      username: 'cdv1x',
      bio: '',
      image: '',
      gender: '',
      first_name: '',
      last_name: '',
      location: '',
      birth_date: null,
    },
    slug: 'string',
    image: 'http://string.com',
    average_rating: 0,
    tagName: [],
  },
  {
    title: 'it tgf friday',
    description: 'day of the week',
    body: "You can't miss this tdf friday in ruby",
    createdAt: '2019-06-20T13:17:46.909484Z',
    updatedAt: '2019-06-20T13:17:46.909533Z',
    author: {
      email: 'alexxsanya@gmail.com',
      username: 'alexxsanya',
      bio: '',
      image: '',
      gender: 'male',
      first_name: 'Alex',
      last_name: 'Sanya',
      location: '',
      birth_date: null,
    },
    slug: 'friday',
    image: 'https://via.placeholder.com/236x124',
    average_rating: 0,
    tagName: [],
  }];
