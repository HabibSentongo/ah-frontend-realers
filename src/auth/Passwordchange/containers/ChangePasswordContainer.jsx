import { connect } from 'react-redux';
import changePasswordAction from '../redux/actions/postAction';

const mapStateToProps = state => ({
  request_success: state.changePassword.success,
  request_errors: state.changePassword.errors,
  isLoading: state.changePassword.isLoading,
});

export default connect(mapStateToProps, { changePasswordAction });
