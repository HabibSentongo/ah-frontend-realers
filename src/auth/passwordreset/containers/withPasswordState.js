import { connect } from 'react-redux';
import resetPasswordAction from '../redux/actions/postAction';

const mapStateToProps = state => ({
  request_success: state.resetPassword.success,
  request_errors: state.resetPassword.errors,
  isLoading: state.resetPassword.isLoading,
});

export default connect(mapStateToProps, { resetPasswordAction });
