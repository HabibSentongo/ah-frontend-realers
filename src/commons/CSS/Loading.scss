@import '../assets/variables/colors.scss';
@import '../assets/variables/font.scss';

.loading {
  left: 47%;
  position: absolute;
  top: 55%;

  .spinner-border {
    width: 5rem;
    height: 5rem;
    border-color:$primary-color;
    vertical-align: text-bottom;
    border-right-color: transparent;
}
}
