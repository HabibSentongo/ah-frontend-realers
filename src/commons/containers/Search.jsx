import {
  Nav, Form, InputGroup, FormControl, Button,
} from 'react-bootstrap';
import React from 'react';
import '../CSS/Search.scss';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faSearch } from '@fortawesome/free-solid-svg-icons';

export class Search extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      searchKey: '',
    };
  }

  handleChange = (event) => {
    const { value } = event.target;
    this.setState({
      searchKey: value,
    });
  }

  render() {
    const { searchKey } = this.state;
    return (
      <Nav className="justify-content-center bg-dark Nav" activeKey="/home">
        <InputGroup className="search-box">
          <InputGroup.Prepend>
            <InputGroup.Text className="search-icon">
              <FontAwesomeIcon icon={faSearch} />
            </InputGroup.Text>
          </InputGroup.Prepend>
          <Form inline onSubmit={(e) => { e.preventDefault(); }}>
            <FormControl
                type="text"
                placeholder="Search articles by tag, title or author ..."
                className="mr-sm-2 search-input"
                onChange={this.handleChange}
                value={searchKey}
                required
              />
            <Button type="button" href={`/search/${searchKey}`} className="search-btn">SEARCH</Button>
          </Form>
        </InputGroup>
      </Nav>
    );
  }
}

export default Search;
