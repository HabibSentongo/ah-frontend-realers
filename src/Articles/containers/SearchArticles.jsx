import React, { Component } from 'react';
import { Container } from 'react-bootstrap';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import Article from '../../commons/components/Article';
import fetchSearchedArticles from '../redux/actions/SearchArticlesAction';
import '../CSS/SearchArticles.scss';
import ErrorLabel from '../../commons/components/ErrorLabel';
import Loading from '../../commons/components/Loading';

export class SearchArticles extends Component {
  componentDidMount() {
    const { match: { params: { searchkey } }, getSearchedArticles } = this.props;

    getSearchedArticles(searchkey);
  }

  render() {
    const { items, match: { params: { searchkey } }, isSearching } = this.props;
    return (
      <div className="search-articles">
        {
        isSearching === true
          ? <Loading />
          : (
            <Container>
              {items.length !== 0 ? (
                <div>
                  <h2 className="section-title">{`Search Results For - '${searchkey}' `}</h2>
                  <div className="article-display-box">
                    {items.map(article => (
                      <Article key={article.slug} article={article} />
                    ))}
                  </div>
                </div>
              ) : (
                <ErrorLabel
                  message={`Your search - '${searchkey}' did not match any of the articles`}
                  tip="Try again with another word"
                />
              )}
            </Container>
          )
        }
      </div>
    );
  }
}

SearchArticles.propTypes = {
  items: PropTypes.array,
  match: PropTypes.object.isRequired,
  isSearching: PropTypes.bool,
  getSearchedArticles: PropTypes.func.isRequired,
};

SearchArticles.defaultProps = {
  items: [],
  isSearching: false,
};

export const mapStateToProps = state => {
  const { isSearching, items } = state.searchedArticles;
  return { isSearching, items };
};

export const mapDispatchToProps = dispatch => ({
  getSearchedArticles: (searchedKey) => {
    dispatch(fetchSearchedArticles(searchedKey));
  },
});

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(SearchArticles);
