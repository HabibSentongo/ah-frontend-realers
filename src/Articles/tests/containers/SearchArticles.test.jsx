/* eslint-disable no-undef */
import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { SearchArticles, mapStateToProps, mapDispatchToProps } from '../../containers/SearchArticles';
import { ARTICLE_DATA } from '../../../__mock__/articleData';
import Loading from '../../../commons/components/Loading';

const props = {
  items: ARTICLE_DATA,
  match: {
    params: {
      searchkey: 'article',
    },
  },
  isSearching: false,
  getSearchedArticles: jest.fn(),
};

describe('SearchArticles Component', () => {
  const component = shallow(<SearchArticles {...props} />);

  it('should match the snapshot', () => {
    expect(toJson(component)).toMatchSnapshot();
  });

  it('getSearchedArticles is called upon mount', () => {
    const wrapper = shallow(<SearchArticles {...props} />);
    wrapper.setProps(props);
    expect(wrapper).toBeDefined();
    expect(props.getSearchedArticles).toHaveBeenCalled();
  });

  it('should display loading before retrieving articles', () => {
    const wrapper = shallow(<SearchArticles {...props} />);
    props.isSearching = true;
    wrapper.setProps(props);
    expect(wrapper.contains(<Loading />)).toBe(true);
  });

  it('should display no available message when no articles returned', () => {
    const wrapper = shallow(<SearchArticles {...props} />);
    props.items = [];
    props.isSearching = false;
    wrapper.setProps(props);
    const ErrorLabel = wrapper.find('ErrorLabel');
    expect(ErrorLabel.prop('tip')).toEqual('Try again with another word');
  });

  it('props should match the mockedstate', () => {
    const mockedState = {
      searchedArticles: {
        items: [],
        status: true,
        isSearching: false,
        message: 'Article Search Successful',
      },
    };
    const state = mapStateToProps(mockedState);
    expect(state.isSearching).toBe(false);
  });

  it('mapStateToProps should return the right value', () => {
    const dispatch = jest.fn();
    mapDispatchToProps(dispatch).getSearchedArticles('hello');
    expect(dispatch).toHaveBeenCalled();
  });
});
