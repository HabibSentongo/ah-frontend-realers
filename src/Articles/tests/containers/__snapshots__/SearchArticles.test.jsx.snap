// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchArticles Component should match the snapshot 1`] = `
<div
  className="search-articles"
>
  <Container
    fluid={false}
  >
    <div>
      <h2
        className="section-title"
      >
        Search Results For - 'article' 
      </h2>
      <div
        className="article-display-box"
      >
        <Article
          article={
            Object {
              "author": Object {
                "bio": "",
                "birth_date": null,
                "email": "cedric.lusiba@andela.com",
                "first_name": "",
                "gender": "",
                "image": "",
                "last_name": "",
                "location": "",
                "username": "cdv1x",
              },
              "average_rating": 0,
              "body": "strinfghjhhg",
              "createdAt": "2019-06-19T09:16:23.749462Z",
              "description": "string",
              "image": "http://string.com",
              "slug": "string",
              "tagName": Array [],
              "title": "string",
              "updatedAt": "2019-06-19T09:16:23.749489Z",
            }
          }
          key="string"
        />
        <Article
          article={
            Object {
              "author": Object {
                "bio": "",
                "birth_date": null,
                "email": "alexxsanya@gmail.com",
                "first_name": "Alex",
                "gender": "male",
                "image": "",
                "last_name": "Sanya",
                "location": "",
                "username": "alexxsanya",
              },
              "average_rating": 0,
              "body": "You can't miss this tdf friday in ruby",
              "createdAt": "2019-06-20T13:17:46.909484Z",
              "description": "day of the week",
              "image": "https://via.placeholder.com/236x124",
              "slug": "friday",
              "tagName": Array [],
              "title": "it tgf friday",
              "updatedAt": "2019-06-20T13:17:46.909533Z",
            }
          }
          key="friday"
        />
      </div>
    </div>
  </Container>
</div>
`;
